src/org/traccar/database/CommandsManager.java

```diff
@@ -71,32 +71,26 @@ public void sendCommand(Command command, long deviceId, boolean fallbackToText)
             if (activeDevice != null) {
                 activeDevice.sendCommand(command);
             } else {
-                if (fallbackToText) {
-                    command.setTextChannel(true);
-                    sendCommand(command, deviceId, false);
-                } else {
-                    throw new RuntimeException("Device is not online");
-                }
+                sent = !getDeviceQueue(deviceId).add(command);
             }
         }
+        return sent;
     }
 
     public Collection<Long> getSupportedCommands(long deviceId) {
         List<Long> result = new ArrayList<>();
         Position lastPosition = Context.getIdentityManager().getLastPosition(deviceId);
-        boolean online = Context.getConnectionManager().getActiveDevice(deviceId) != null;
         for (long commandId : getAllDeviceItems(deviceId)) {
             Command command = getById(commandId);
-            if (command.getTextChannel() || online) {
-                if (lastPosition != null) {
-                    BaseProtocol protocol = Context.getServerManager().getProtocol(lastPosition.getProtocol());
-                    if (protocol.getSupportedTextCommands().contains(command.getType())
-                            || online && protocol.getSupportedDataCommands().contains(command.getType())) {
-                        result.add(commandId);
-                    }
-                } else if (command.getType().equals(Command.TYPE_CUSTOM)) {
+            if (lastPosition != null) {
+                BaseProtocol protocol = Context.getServerManager().getProtocol(lastPosition.getProtocol());
+                if ((command.getTextChannel()
```

Comment:
This construction might looks unclear, can unwrap ternary operator.

Reply:
Just noticed this one. Please unwrap it.